### PR TITLE
fix(backend): stop a redelivered part payment retrying forever

### PR DESCRIPTION
**Pull request by ankit-yc**
## What changed

`FinancePaymentService.recordInvoicePayment` now recovers from its own unique violation on the `Payment` insert, returning the payment that is already recorded instead of throwing.

## Why

A redelivered Stripe payment event for a **deposit or a partial payment** throws an uncaught P2002 today. The endpoint answers non-2xx, Stripe retries, and the retry hits the same wall, so the event is retried until Stripe gives up. Three things line up to produce it:

- the `ALREADY_PAID` guard only fires once the invoice is `PAID`, and `updateInvoiceAfterPayment` only sets `PAID` when the applied amount covers the **whole** balance (`if (appliedAmount >= balance)`), so a part payment never trips it;
- the attempt lookup excludes only `CANCELED` and `FAILED`, so it finds the now-`SUCCEEDED` attempt from the first delivery and passes its id along;
- `recordInvoicePayment` then inserts a `Payment` carrying that attempt id, and `Payment.paymentAttemptId` is `@unique`.

There is no transaction around any of it, so the attempt row has already been mutated by the time the insert dies.

The recovery deliberately does **not** fall through to `updateInvoiceAfterPayment`. The money was applied by the delivery that won; applying it again would double-count it against the invoice. `appliedAmount` comes back as `0` for the same reason, so a caller accumulating it cannot count the same payment twice.

## How this was found

While checking whether `PaymentAttempt.providerPaymentIntentId` could take a unique constraint, per the back-office plan. A five-angle safety review concluded the constraint **is** safe but **would not have fixed anything**: a redelivered webhook UPDATEs the attempt on a row selected BY that same intent id, so the index is never consulted. This is the failure that was actually reachable, so it is what this PR fixes. The constraint is not added here.

Two things worth recording for whoever picks this up next:

- `PaymentAttemptStatus.FAILED` is **never written anywhere** in the codebase. `handleInvoicePaymentFailed` marks the *invoice* FAILED, not the attempt. Any reasoning that treats "no FAILED attempts exist" as evidence about an untested path is wrong; it is a property of the code.
- Nothing collides on the intent id today only because `stripe.paymentIntents.create` is called with **no idempotency key**, so Stripe returns a fresh id every time. Adding one, which looks like ordinary hardening, would make a retried create return the same id. If the unique on `PaymentAttempt.providerPaymentIntentId` is ever added, that call site needs a comment saying so, or the customer-facing pay-now request becomes a P2002.

## Impact area

`apps/backend` only. No schema change, no migration, no client change.

## Validation performed

- `pnpm --filter backend run type-check` - clean
- `pnpm --filter backend run lint` - exit 0, no findings
- `npx jest test/services/finance.payment.test.ts` - **119 passed**
- `npx prettier --check` on both changed files - clean

Both behaviours are mutation-verified, and each mutation fails a different test:

| mutation | result |
|---|---|
| remove the P2002 catch (the bug exactly as it stands on `dev`) | `● treats a redelivered partial payment as a replay instead of throwing` |
| report `appliedAmount` instead of `0` (the double-count) | `● returns a no-op result when the invoice is already fully paid` - an **existing** test |

The second one is the useful signal: an invariant that was already in the suite catches the double-count, so the guard cannot be loosened into re-applying money without something going red.

### CODE REVIEW

File: apps/backend/src/services/finance/payment.ts [hunk 1]
```diff
@@ -1838,22 +1845,69 @@ export const FinancePaymentService = {
...
+
+      logger.info(
+        `Payment for attempt ${paymentAttempt.id} was already recorded; treating this delivery as a replay`,
+      );
+
+      return {
+        invoice: currentInvoice,
+        paymentAttempt,
+        payment: existingPayment,
+        balanceAfterPayment: settled.balance,
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/backend/src/services/finance/payment.ts]: Real, and the sharper of the two. Fixed in 94950a614.

You are right that the P2002 proves only that the Payment row exists. There is no transaction around the insert and the invoice update that follows it, so the delivery that won can die in between, and my first version would then have answered 200 and ended the retries that were its last chance to finish. A fully covered invoice could have stayed not PAID permanently. That is worse than the bug I was fixing, because the retry loop at least kept trying.

I took the second option you offered rather than atomicity, because making the insert and its side effects atomic means threading a transaction client through markInvoiceTreatmentItemsSettled in ./settlement, and that is a larger change to money code than this PR should carry. Worth doing separately.

The recovery path now re-derives the position instead of assuming it. It re-reads the invoice with its SUCCEEDED payments, so the balance already accounts for the winner payment, and if the invoice is covered but not PAID it completes that settlement and the treatment items. Idempotent by construction: an invoice already PAID skips it.

What this does NOT reconcile, stated plainly rather than left implicit: depositCollectedAmount is a stored accumulator incremented per payment, with no marker saying whether a given payment was counted, so a delivery that died between the insert and the update leaves it stale and no read can tell. Settling a covered invoice is the money-visible case and this closes it; the deposit accumulator needs either the atomicity fix above or a derivation from the payments, and I did not want to guess at that here.

Mutation-checked: forcing needsSettling to false fails the new test "settles an invoice the winning delivery paid for but never marked PAID" and nothing else.

File: apps/backend/src/services/finance/payment.ts [hunk 2]
```diff
@@ -1838,22 +1845,69 @@ export const FinancePaymentService = {
...
+        paidToDate: settled.paid,
+        // Nothing new was applied. A caller that adds this to a running total
+        // must not count the same money twice.
+        appliedAmount: 0,
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/backend/src/services/finance/payment.ts]: Also real. Fixed in the same commit.

The replay path returned a shape that spreads into action PAID, and _handleInvoiceCheckout notifies the parent for every PAID result with a parentId. Since Stripe redelivers whenever it does not receive a 2xx, including when the response was simply lost, a parent whose payment already succeeded and who was already notified would be told a second time.

Both return paths now carry an explicit boolean flag rather than relying on appliedAmount being zero, because a genuine zero-amount application is conceivable and the distinction being tested here is "was anything newly applied", which deserves to be said outright. The checkout handler returns early when it is set.

Mutation-checked: ignoring the flag fails the new test "does not notify the parent again when the checkout session is replayed" and nothing else.

File: apps/backend/src/services/finance/payment.ts [hunk 3]
```diff
@@ -1838,22 +1845,103 @@ export const FinancePaymentService = {
...
-        rawProviderPayload: input.rawProviderPayload ?? undefined,
-      },
-    });
+        currentInvoice.totalAmount,
+        currentInvoice.depositCollectedAmount ?? 0,
+      );
+
+      const needsSettling =
+        settled.balance <= 0 && currentInvoice.status !== "PAID";
+      if (needsSettling) {
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/backend/src/services/finance/payment.ts]: Correct, and this is the one that made me stop patching. Fixed in 36c7680be, by removing the cause rather than adding a fourth guard.

You are right on the mechanism: the balance <= 0 early return fires before the insert is ever attempted, so a redelivery after a full-balance payment never reached the recovery I added, and the invoice stayed AWAITING_PAYMENT. You are also right that the test could not represent it, because it mocked the retry initial payment lookup as empty, which cannot be the state after the winner insert. That was a badly constructed test and it is gone.

The root cause is that the attempt write, the Payment insert and the invoice update were three sequential awaits with no transaction. Every guard built on top of that was unsound in a different way, and no reasoning on the recovery path could distinguish them, because nothing records which of the three steps ran. They are now inside one interactive transaction, so the Payment row existing genuinely implies the invoice moved with it. That is what makes returning what is recorded a safe answer instead of an assumption, and it removes this scenario rather than handling it.

markInvoiceTreatmentItemsSettled and createPaymentAttempt now take the client so their writes join the transaction. FinanceEventService.recordEvent stays outside on purpose: it closes over the module-level client, so calling it inside would run on a second connection against rows the transaction still holds.

On the notification half of your point: the early return now carries the replay flag too, so a delivery arriving after the balance is already closed no longer reads as a fresh success. That branch initially passed with and without the flag, so I added a test rather than trust it.

Not covered, deliberately: invoices already left inconsistent by the old non-atomic path are not repaired. Forward fix; repairing existing rows is a data task.

[the remark this answers is not in the document]
Comment (ankit-yc) [reply apps/backend/src/services/finance/payment.ts]: Also correct, and resolved by the same change in 36c7680be rather than by special-casing deposits.

Your point was that a deposit delivery which creates its Payment but fails before updateInvoiceAfterPayment leaves depositCollectedAmount and billingCollectionMode stale while the payment row records the money, and that my recovery only touched fully covered invoices so it would not restore them.

I had flagged that as unfixable from the recovery path, and that was the right diagnosis of the symptom but the wrong conclusion. depositCollectedAmount is an accumulator with no marker saying whether a given payment was counted, so no read can tell whether the dead delivery applied it. The answer is not to reconstruct it afterwards, it is for the gap never to exist: the deposit branch of updateInvoiceAfterPayment now runs inside the same transaction as the Payment insert, so either both happened or neither did.

That is why this went to atomicity rather than a third patch. Two of your three findings on this PR were different symptoms of the same missing transaction, and the third, the early return, was a fourth.
